### DISCUSSION TITLE
Kotlin Multiplatform Use Cases Survey

### DISCUSSION THREAD

**Discussion by aajtodd** [aws/aws-sdk-kotlin#464] [General]
We would like to hear from you about your use cases for a multiplatform AWS SDK for Kotlin.

In particular we are interested in hearing about what target(s) you need support for and what you would like to do with the SDK (e.g. cross platform mobile (KMM), desktop, backend server, etc).

These insights are invaluable to us better understanding your needs as well as prioritizing support for multiplatform targets.

Example responses we are looking for 

-----------

**KMP Targets**: `jvm`, `linuxX64`
**Use**: backend server development (lambda and ecs)

-----------

**KMP Targets**: `android`, `ios`
**Use**: KMM

-----------

NOTE:
* Please list KMP targets using their [target preset name ](https://kotlinlang.org/docs/mpp-supported-platforms.html)
* Feel free to add any other relevant details you wish to share
* Be sure to upvote 👍  existing issues for [Kotlin/Native](https://github.com/awslabs/aws-sdk-kotlin/issues/229) and [Kotlin/JS](https://github.com/awslabs/aws-sdk-kotlin/issues/228)!

**Comment by napperley** [comment]
----
**KMP Targets**: `jvm`
**Use**: Backend server development (Ec2)
----

----
**KMP Targets**: `js`
**Use**: Web client development
----

----
**KMP Targets**: `linuxX64`, `linuxArm32Hfp`, `linuxArm64`
**Use**: Backend service development (Ec2), Serverless development (Lambda), Desktop development, High level Linux/Embedded Linux development, CLI client development
----

One key point of interest is in having standard communication methods between a Kotlin client app (running on a local/edge device), and a Kotlin server app (running on a AWS Ec2 Micro instance for example). Increasingly it is becoming more common to find Kotlin projects that are full stack based (eg desktop app for frontend, and server app for backend). Sharing of backend logic and data models is very important for full stack projects.

Currently a Kotlin Native program is running in production (as a [Linux Daemon](https://bash.cyberciti.biz/guide/Daemons#:~:text=A%20daemon%20(also%20known%20as,start%20daemons%20at%20boot%20time.)) on a AWS Ec2 Micro instance via the **linuxX64** target) that handles IoT data processing/analytics. All of the program's workloads are I/O bound.

**Comment by markth0mas** [comment]
I use both Kotlin JS and Kotlin JVM for lambdas, and have a few multi platform libraries that I share between them. Being able to share code that calls AWS APIs would make some of those libraries easier to implement.

**Comment by drofwarcs** [comment]
KMP Targets: android, ios, jvm, js
Use: KMM/KMP

Working on KMM project that would benefit from easily integrating with some of AWS services. Even though starting out with mobile targets would be the most in demand, I believe that extending support to js and jvm would be just as beneficial. Was looking into using Amplify as option for mBaas and think this would be an awesome usecase for AWS to integrate KMP. Not sure if it will require the dependent services (cognito, appsync, S3...etc) to also have KMP support, but a good first start would be to have some way for Amplify to be KMP compatible and delegate to the native sdks for each indipendent service, similar to the way Gitlive did for their [firebase KMP wrapper](https://firebaseopensource.com/projects/gitliveapp/firebase-kotlin-sdk/) .

**Comment by Sianay** [comment]
KMP Targets: android, ios
Use: KMM

Would like to be able to use AWS API in a KMM project

**Comment by felixbaral** [comment]
I share a library between native android and ios clients and just need to download files from my S3 Buckets.

**Comment by toufeeqahamedns** [comment]
KMP Targets: android, ios
Use: KMM

Using it on mobile platforms would be great use case

**Comment by maggie-weber** [comment]
KMP Targets: `jvm`, `js`
Use: An internal library/SDK for our web application that can be used both by our frontend HTML/JS and a desktop application. Uploading / Downloading with S3.

**Comment by matpag** [comment]
KMP Targets: android, ios
Use: KMM

**Comment by brianguertin** [comment]
KMP Targets: android, ios, js
Use: KMM

We ship several dozen native mobile and web apps that share a common Kotlin code base. A multiplatform SDK would let us integrate features from the Amazon SDK in the "common" source code, instead of implementing it 3 times (one for each platform).

**Comment by drakon64** [comment]
KMP Targets: js
Use: backend server development (lambda)

Compiling Kotlin to JavaScript rather than JVM reduces the cold-start time for Lambda functions

**Comment by gsteckman** [comment]
KMP Targets: linuxX64
Use: backend server development (lambda)

Compiling to native rather than JVM reduces cold-start time and memory footprint for Lambda functions.
